src/components/Table/utils/RowStates.ts

```diff
@@ -0,0 +1,110 @@
+import { ObservableMap } from "mobx";
+import { GridDataRow, HEADER, KEPT_GROUP, reservedRowKinds } from "src";
+import { RowState } from "src/components/Table/utils/RowState";
+
+/**
+ * Manages our tree of observable RowStates that manage each GridDataRow's behavior.
+ */
+export class RowStates {
+  // A flat map of all row id -> RowState
+  private map = new ObservableMap<string, RowState>();
+  // Pre-create our keptGroupRow for if/when we need it.
+  keptGroupRow: RowState = creatKeptGroupRow();
+
+  /** Returns a flat list of all of our RowStates. */
+  get allStates(): RowState[] {
+    return [...this.map.values()];
+  }
+
+  /** Returns the `RowState` for the given `id`. We should probably require `kind`. */
+  get(id: string): RowState {
+    const rs = this.map.get(id);
+    if (!rs) throw new Error(`No RowState for ${id}`);
+    return rs;
+  }
+
+  /**
+   * Merge a new set of `rows` prop into our state.
+   *
+   * Any missing rows are marked as `wasRemoved` so we can consider them "kept" if they're also selected.
+   */
+  setRows(rows: GridDataRow<any>[]): void {
+    const existing = new Set(this.map.values());
+    const map = this.map;
+
+    function addRowAndChildren(parent: RowState | undefined, row: GridDataRow<any>): RowState {
+      // This should really be kind+id, but a lot of our lookups just use id
+      const key = row.id;
+      let state = map.get(key);
+      if (!state) {
+        state = new RowState(parent, row);
+        map.set(key, state);
+      } else {
+        state.parent = parent;
+        state.row = row;
+        state.wasRemoved = false;
+        existing.delete(state);
+      }
+      state.children = row.children?.map((child) => addRowAndChildren(state, child));
+      return state;
+    }
+
+    // Probe for the header row, so we can create it as a root RowState, even
+    // though we don't require the user to model their GridDataRows that way.
+    const header = rows.find((r) => r.kind === HEADER);
```

Comment:
**thought:** We shouldn't have this as an issue right now, but I know it's been asked before... what if we have multiple `kind === "header"`? Clearly right now we only support a single `header` kind. And I think we should continue to do that. But thinking that maybe we should warn the implementer somehow if we find multiple `header`s. Or they'll just see that its not working as expected 🤷

Reply:
Yeah; I'm tempted to think that the "first header" will always be important/in control of things like select-all/collapse-all; so if they had two headers, I think that would technically be fine, it's just that the 2nd one wouldn't be able to drive any special select-all behavior.

If they wanted that to happen, it seems like the "2nd header" would be "the real header" and we'd use some other kind, like totals, to be a psuedo-header that comes first...